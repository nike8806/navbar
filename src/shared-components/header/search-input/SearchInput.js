import React, { Component } from 'react';
import Autosuggest from 'react-autosuggest';
import axios from 'axios';
import PropTypes from 'prop-types';

import './SearchInput.scss';

const { CancelToken } = axios;
const apiJobs = 'https://api.dataatwork.org/v1/';
const jobsUrls = {
  autocomplete: 'jobs/autocomplete',
  detail: 'jobs/'
};

const getSuggestionValue = (suggestion) => {
  axios.get(apiJobs + jobsUrls.detail + suggestion.uuid).then((response) => {
    console.log(response.data);
  }).catch((error) => {
    console.log(error);
  });
  return suggestion.suggestion;
};

// Use your imagination to render suggestions.
const renderSuggestion = suggestion => (
  <div>
    {suggestion.suggestion}
  </div>
);


/**
 * Handle the Search Input page
 */
class Header extends Component {
  static propTypes = {
    classNameComponent: PropTypes.string.isRequired
  };

  constructor() {
    super();
    this.state = {
      value: '',
      suggestions: []
    };
  }

  onChange = (event, { newValue }) => {
    this.setState({
      value: newValue
    });
  };

  loadSuggestions = (value) => {
    if (value.length < 2) {
      return;
    }

    // Cancel the previous request
    if (this.lastRequestId) {
      this.lastRequestId.cancel('Operation canceled by the user.');
    }

    this.setState({
      isLoading: true
    });

    const cancelTokenSource = CancelToken.source();
    this.lastRequestId = cancelTokenSource;

    axios.get(apiJobs + jobsUrls.autocomplete, {
      params: {
        begins_with: value
      },
      cancelToken: cancelTokenSource.token
    }).then((response) => {
      const suggestions = response.data.slice(0, 10);
      this.setState({
        suggestions,
        isLoading: false
      });
    }).catch((error) => {
      console.log(error);
      this.setState({
        suggestions: [],
        isLoading: false
      });
    });
  }

  submitRequestSuggestions = (e) => {
    e.preventDefault();
    const autocomplete = e.target.elements.item('input.search_input');
    if (autocomplete) {
      this.loadSuggestions(autocomplete.value);
    }
  }

  // Autosuggest will call this function every time you need to update suggestions.
  // You already implemented this logic above, so just use it.
  onSuggestionsFetchRequested = ({ value }) => {
    this.loadSuggestions(value);
  };

  // Autosuggest will call this function every time you need to clear suggestions.
  onSuggestionsClearRequested = () => {
    this.setState({
      suggestions: []
    });
  };

  onFocusInput = (e) => {
    e.target.classList.add('search__input--large');
  }

  onBlurInput = (e) => {
    e.target.classList.remove('search__input--large');
  }

  render() {
    const { classNameComponent } = this.props;
    const { value, suggestions, isLoading } = this.state;

    // Autosuggest will pass through all these props to the input.
    const inputProps = {
      placeholder: 'Type a Job',
      value,
      onChange: this.onChange,
      className: 'search__input',
      name: 'autocomplete',
      onFocus: this.onFocusInput,
      onBlur: this.onBlurInput
    };

    return (
      <form
        className={`search ${classNameComponent}`}
        onSubmit={this.submitRequestSuggestions}
      >
        <Autosuggest
          suggestions={suggestions}
          onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
          onSuggestionsClearRequested={this.onSuggestionsClearRequested}
          getSuggestionValue={getSuggestionValue}
          renderSuggestion={renderSuggestion}
          inputProps={inputProps}
        />
        <button type="submit">
          <i className={`fas ${!isLoading ? 'fa-search' : 'fa-spinner'}`} />
        </button>
      </form>
    );
  }
}

export default Header;
