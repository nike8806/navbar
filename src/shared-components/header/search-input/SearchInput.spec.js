import React from 'react';
import { shallow } from 'enzyme';
import axios from 'axios';

import SearchInput from './SearchInput';


let component;

describe.only('SearchInput component', () => {
  describe('On mount', () => {
    let searchInput;
    beforeAll(() => {
      component = shallow(
        <SearchInput classNameComponent="header__widgets__search" />
      );
    });

    it('should render correctly', () => {
      expect(component).toMatchSnapshot();
    });

    describe('On content changed', () => {
      // const handleChangeSpy = sinon.spy(New.prototype, "handleChange");
      // let instance;
      // let handleChangeSpy;
      const event = { arget: { value: 'any' } };
      beforeEach(() => {
        searchInput = component.find('form');
      });

      it('Should call to get service when the input value has more than 1 caracter', async () => {
        searchInput.simulate('onSubmit', event);
        // axios.get = jest.fn().mockResolvedValue({});
        // expect(axios.get).toBeCalled();
        // expect(axios.get).toBeCalledWith('/todo/kba/questions');
      });
    });

    describe('On focus', () => {

    });
    describe('On blur', () => {

    });
  });
});
