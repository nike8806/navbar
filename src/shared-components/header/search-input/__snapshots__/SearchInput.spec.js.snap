// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput component On mount should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Header
    classNameComponent="header__widgets__search"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Autosuggest
          alwaysRenderSuggestions={false}
          focusInputOnSuggestionClick={true}
          getSuggestionValue={[Function]}
          highlightFirstSuggestion={false}
          id="1"
          inputProps={
            Object {
              "className": "search__input",
              "name": "autocomplete",
              "onBlur": [Function],
              "onChange": [Function],
              "onFocus": [Function],
              "placeholder": "Type a Job",
              "value": "",
            }
          }
          multiSection={false}
          onSuggestionsClearRequested={[Function]}
          onSuggestionsFetchRequested={[Function]}
          renderSuggestion={[Function]}
          renderSuggestionsContainer={[Function]}
          shouldRenderSuggestions={[Function]}
          suggestions={Array []}
          theme={
            Object {
              "container": "react-autosuggest__container",
              "containerOpen": "react-autosuggest__container--open",
              "input": "react-autosuggest__input",
              "inputFocused": "react-autosuggest__input--focused",
              "inputOpen": "react-autosuggest__input--open",
              "sectionContainer": "react-autosuggest__section-container",
              "sectionContainerFirst": "react-autosuggest__section-container--first",
              "sectionTitle": "react-autosuggest__section-title",
              "suggestion": "react-autosuggest__suggestion",
              "suggestionFirst": "react-autosuggest__suggestion--first",
              "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
              "suggestionsContainer": "react-autosuggest__suggestions-container",
              "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
              "suggestionsList": "react-autosuggest__suggestions-list",
            }
          }
        />,
        <button
          type="submit"
        >
          <i
            className="fas fa-search"
          />
        </button>,
      ],
      "className": "search header__widgets__search",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "alwaysRenderSuggestions": false,
          "focusInputOnSuggestionClick": true,
          "getSuggestionValue": [Function],
          "highlightFirstSuggestion": false,
          "id": "1",
          "inputProps": Object {
            "className": "search__input",
            "name": "autocomplete",
            "onBlur": [Function],
            "onChange": [Function],
            "onFocus": [Function],
            "placeholder": "Type a Job",
            "value": "",
          },
          "multiSection": false,
          "onSuggestionsClearRequested": [Function],
          "onSuggestionsFetchRequested": [Function],
          "renderSuggestion": [Function],
          "renderSuggestionsContainer": [Function],
          "shouldRenderSuggestions": [Function],
          "suggestions": Array [],
          "theme": Object {
            "container": "react-autosuggest__container",
            "containerOpen": "react-autosuggest__container--open",
            "input": "react-autosuggest__input",
            "inputFocused": "react-autosuggest__input--focused",
            "inputOpen": "react-autosuggest__input--open",
            "sectionContainer": "react-autosuggest__section-container",
            "sectionContainerFirst": "react-autosuggest__section-container--first",
            "sectionTitle": "react-autosuggest__section-title",
            "suggestion": "react-autosuggest__suggestion",
            "suggestionFirst": "react-autosuggest__suggestion--first",
            "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
            "suggestionsContainer": "react-autosuggest__suggestions-container",
            "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
            "suggestionsList": "react-autosuggest__suggestions-list",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <i
            className="fas fa-search"
          />,
          "type": "submit",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "fas fa-search",
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Autosuggest
            alwaysRenderSuggestions={false}
            focusInputOnSuggestionClick={true}
            getSuggestionValue={[Function]}
            highlightFirstSuggestion={false}
            id="1"
            inputProps={
              Object {
                "className": "search__input",
                "name": "autocomplete",
                "onBlur": [Function],
                "onChange": [Function],
                "onFocus": [Function],
                "placeholder": "Type a Job",
                "value": "",
              }
            }
            multiSection={false}
            onSuggestionsClearRequested={[Function]}
            onSuggestionsFetchRequested={[Function]}
            renderSuggestion={[Function]}
            renderSuggestionsContainer={[Function]}
            shouldRenderSuggestions={[Function]}
            suggestions={Array []}
            theme={
              Object {
                "container": "react-autosuggest__container",
                "containerOpen": "react-autosuggest__container--open",
                "input": "react-autosuggest__input",
                "inputFocused": "react-autosuggest__input--focused",
                "inputOpen": "react-autosuggest__input--open",
                "sectionContainer": "react-autosuggest__section-container",
                "sectionContainerFirst": "react-autosuggest__section-container--first",
                "sectionTitle": "react-autosuggest__section-title",
                "suggestion": "react-autosuggest__suggestion",
                "suggestionFirst": "react-autosuggest__suggestion--first",
                "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
                "suggestionsContainer": "react-autosuggest__suggestions-container",
                "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
                "suggestionsList": "react-autosuggest__suggestions-list",
              }
            }
          />,
          <button
            type="submit"
          >
            <i
              className="fas fa-search"
            />
          </button>,
        ],
        "className": "search header__widgets__search",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "alwaysRenderSuggestions": false,
            "focusInputOnSuggestionClick": true,
            "getSuggestionValue": [Function],
            "highlightFirstSuggestion": false,
            "id": "1",
            "inputProps": Object {
              "className": "search__input",
              "name": "autocomplete",
              "onBlur": [Function],
              "onChange": [Function],
              "onFocus": [Function],
              "placeholder": "Type a Job",
              "value": "",
            },
            "multiSection": false,
            "onSuggestionsClearRequested": [Function],
            "onSuggestionsFetchRequested": [Function],
            "renderSuggestion": [Function],
            "renderSuggestionsContainer": [Function],
            "shouldRenderSuggestions": [Function],
            "suggestions": Array [],
            "theme": Object {
              "container": "react-autosuggest__container",
              "containerOpen": "react-autosuggest__container--open",
              "input": "react-autosuggest__input",
              "inputFocused": "react-autosuggest__input--focused",
              "inputOpen": "react-autosuggest__input--open",
              "sectionContainer": "react-autosuggest__section-container",
              "sectionContainerFirst": "react-autosuggest__section-container--first",
              "sectionTitle": "react-autosuggest__section-title",
              "suggestion": "react-autosuggest__suggestion",
              "suggestionFirst": "react-autosuggest__suggestion--first",
              "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
              "suggestionsContainer": "react-autosuggest__suggestions-container",
              "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
              "suggestionsList": "react-autosuggest__suggestions-list",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <i
              className="fas fa-search"
            />,
            "type": "submit",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "fas fa-search",
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
